tests/federation/test_federation_client.py

```diff
@@ -0,0 +1,147 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import json
+from unittest import mock
+
+import twisted.web.client
+from twisted.internet import defer
+from twisted.internet.protocol import Protocol
+from twisted.python.failure import Failure
+from twisted.test.proto_helpers import MemoryReactor
+
+from synapse.api.room_versions import RoomVersions
+from synapse.server import HomeServer
+from synapse.types import JsonDict
+from synapse.util import Clock
+
+from tests.unittest import FederatingHomeserverTestCase
+
+
+class FederationClientTest(FederatingHomeserverTestCase):
+    def prepare(self, reactor: MemoryReactor, clock: Clock, homeserver: HomeServer):
+        super().prepare(reactor, clock, homeserver)
+
+        # mock out the Agent used by the federation client, which is easier than
+        # catching the HTTPS connection and do the TLS stuff.
+        self._mock_agent = mock.create_autospec(twisted.web.client.Agent, spec_set=True)
+        homeserver.get_federation_http_client().agent = self._mock_agent
+
+    def test_get_room_state(self):
+        creator = f"@creator:{self.OTHER_SERVER_NAME}"
+        test_room_id = "!room_id"
+
+        # mock up some events to use in the response
+        create_event_dict = self.add_hashes_and_signatures(
+            {
+                "room_id": test_room_id,
+                "type": "m.room.create",
+                "state_key": "",
+                "sender": creator,
+                "content": {"creator": creator},
+                "prev_events": [],
+                "auth_events": [],
+                "origin_server_ts": 500,
+            }
+        )
+        member_event_dict = self.add_hashes_and_signatures(
+            {
+                "room_id": test_room_id,
+                "type": "m.room.member",
+                "sender": creator,
+                "state_key": creator,
+                "content": {"membership": "join"},
+                "prev_events": [],
+                "auth_events": [],
+                "origin_server_ts": 600,
+            }
+        )
+        pl_event_dict = self.add_hashes_and_signatures(
+            {
+                "room_id": test_room_id,
+                "type": "m.room.power_levels",
+                "sender": creator,
+                "state_key": "",
+                "content": {},
+                "prev_events": [],
+                "auth_events": [],
+                "origin_server_ts": 700,
+            }
+        )
+
+        # mock up the response, and have the agent return it
+        self._mock_agent.request.return_value = defer.succeed(
+            _mock_response(
+                {
+                    "pdus": [
+                        create_event_dict,
+                        member_event_dict,
+                        pl_event_dict,
```

Comment:
I suppose this somewhat goes back to my earlier point of confusion: I would expect the latter 2 events of these to depend on the first in some way, but the code you've written so far claims that the order doesn't matter.

Supposing my expectation was right: that these events should be interdependent in some way, how does it make sense for us to process them in 'any' order where we might come across some before the events they depend on?
I realise we have a concept of 'outliers' but I think I'd benefit from better seeing how this fits together.

Reply:
The thing is that `FederationClient.get_room_state` makes it the *caller's* problem to deal with the interdependencies, which is why you aren't seeing any code here which cares about this.

If it helps put it into context, #12040 builds on this PR with a call to `get_room_state` at https://github.com/matrix-org/synapse/pull/12040/files#diff-a91beeaf44154c199a11f5539f2358704a1df966ff74ad8f2681b64430843f8cR968. It just passes the results into `_auth_and_persist_outliers` (https://github.com/matrix-org/synapse/blob/develop/synapse/handlers/federation_event.py#L1225) which handles the ordering problem by making sure that we persist each event's auth_events before the event itself.

Reply:
Aha, thank you :)